### ISSUE TITLE
Scheduler race condition resulting in skipped versions (mainly with `version: every`)

### ISSUE THREAD

**Issue by vito** [concourse/concourse#2895]
## Bug Report

This `testflight` test fails, if you run it enough times:

https://github.com/concourse/concourse/blob/c3896f1ecd34d1b340d876997c4e4c46b75c9c29/testflight/resource_version_test.go#L45

Here's a CI build showing the failure: https://ci.concourse-ci.org/teams/main/pipelines/concourse/jobs/testflight/builds/273

## Steps to Reproduce

```sh
cd concourse
docker-compose up --build
cd testflight
ginkgo -v -untilItFails -focus='runs builds with every version'
```

## Expected Results

The test should succeed reliably.

## Actual Results

I've seen it fail in under 100 runs locally. I also hacked the `AfterEach` to leave the pipeline around rather than destroying it, and when investigating I found some interesting results.

Build 1:

![screencapture-localhost-8080-teams-testflight-pipelines-tf-pipeline-1-bc653c72-d5e4-406f-5ea5-b8c82697ca6b-jobs-some-passing-job-builds-1-2018-11-30-14_39_27](https://user-images.githubusercontent.com/1880/49311036-0890e300-f4ae-11e8-9bd2-a09547655020.png)

Build 2:

![screencapture-localhost-8080-teams-testflight-pipelines-tf-pipeline-1-bc653c72-d5e4-406f-5ea5-b8c82697ca6b-jobs-some-passing-job-builds-2-2018-11-30-14_39_40](https://user-images.githubusercontent.com/1880/49311069-0d559700-f4ae-11e8-8756-54220df26a8c.png)

Build 3:

![screencapture-localhost-8080-teams-testflight-pipelines-tf-pipeline-1-bc653c72-d5e4-406f-5ea5-b8c82697ca6b-jobs-some-passing-job-builds-3-2018-11-30-14_39_51](https://user-images.githubusercontent.com/1880/49311073-10e91e00-f4ae-11e8-815a-a7320f89ea80.png)

This shows that build 1 ran with `-guid1`, build 2 ran with `-guid2`, but then build 3 ran with...`-guid4`!

**HOWEVER**, if you look at the resource version history, it tells a different story!

![screencapture-localhost-8080-teams-testflight-pipelines-tf-pipeline-1-bc653c72-d5e4-406f-5ea5-b8c82697ca6b-resources-some-resource-2018-11-30-14_40_22](https://user-images.githubusercontent.com/1880/49311099-265e4800-f4ae-11e8-9dd3-6d6e894e0b99.png)

This indicates that build 2 actually recorded `-guid3` as its input! This, I think, is where the bug begins; with `-guid3` recorded as an input, it makes sense for the third build to have "skipped" ahead to `-guid4`.

So why was `-guid3` recorded as an input when the build actually ran with `-guid2`?

I dunno. Figure it out!

## Version Info

* Concourse version: `c3896f1ecd34d1b340d876997c4e4c46b75c9c29`
* Deployment type (BOSH/Docker/binary): Docker
* Infrastructure/IaaS: local
* Browser (if applicable): n/a
* Did this used to work? not sure, wouldn't be surprised if this has been a bug for a while

**Comment by vito** [comment]
Think I figured this out...

This happens when periodic scheduling fires in the middle of the immediate scheduling via manual trigger.

Let's say P is the periodic scheduling thread and M is the trigger. It happens when:

* M creates a job build: https://github.com/concourse/concourse/blob/8eafcb85303bbac6ab0f6a9a2f2e7a801db96181/atc/scheduler/scheduler.go#L111

* P loads all pending builds: https://github.com/concourse/concourse/blob/8eafcb85303bbac6ab0f6a9a2f2e7a801db96181/atc/scheduler/scheduler.go#L46

* M and schedules all pending builds: https://github.com/concourse/concourse/blob/8eafcb85303bbac6ab0f6a9a2f2e7a801db96181/atc/scheduler/scheduler.go#L128

* P does the same: https://github.com/concourse/concourse/blob/8eafcb85303bbac6ab0f6a9a2f2e7a801db96181/atc/scheduler/scheduler.go#L59

* M sees that it's a manually triggered build and resolves the inputs, determining `guid2` as the input: https://github.com/concourse/concourse/blob/8eafcb85303bbac6ab0f6a9a2f2e7a801db96181/atc/scheduler/buildstarter.go#L98

* M saves `guid2` as an input: https://github.com/concourse/concourse/blob/8eafcb85303bbac6ab0f6a9a2f2e7a801db96181/atc/scheduler/buildstarter.go#L163

* P does the same, loading the `VersionsDB` after M has saved the inputs: https://github.com/concourse/concourse/blob/8eafcb85303bbac6ab0f6a9a2f2e7a801db96181/atc/scheduler/buildstarter.go#L112
  * Critically, this dataset now has `guid2` recorded as an input already, to the very same build that's being scheduled.

* Because `guid2` now appears as a prior input already, P resolves the next input as `guid3` instead: https://github.com/concourse/concourse/blob/8eafcb85303bbac6ab0f6a9a2f2e7a801db96181/atc/scheduler/buildstarter.go#L118

* M creates a build plan with `guid2` as the resource version to fetch: https://github.com/concourse/concourse/blob/8eafcb85303bbac6ab0f6a9a2f2e7a801db96181/atc/scheduler/buildstarter.go#L178

* M calls `Resume`, thereby starting the build with `guid2` as the version to fetch: https://github.com/concourse/concourse/blob/8eafcb85303bbac6ab0f6a9a2f2e7a801db96181/atc/scheduler/buildstarter.go#L196

* P saves `guid3` as the build inputs: https://github.com/concourse/concourse/blob/8eafcb85303bbac6ab0f6a9a2f2e7a801db96181/atc/scheduler/buildstarter.go#L163
  * This effectively clears out the `guid2` input and replaces it with `guid3`, because `UseInputs` deletes any existing inputs for the build (i.e. the ones previously recorded by M): https://github.com/concourse/concourse/blob/8eafcb85303bbac6ab0f6a9a2f2e7a801db96181/atc/db/build.go#L852-L855

* P creates a build plan with `guid3` as the resource version to fetch, however and also calls `Resume`: https://github.com/concourse/concourse/blob/8eafcb85303bbac6ab0f6a9a2f2e7a801db96181/atc/scheduler/buildstarter.go#L196
  * This effectively does nothing because the build is already running by this point, though technically it may be a race as to who acquires the lock first.

So...we've got problems. There are parts of the scheduler that fire concurrently (due to manual triggers) but are not thread-safe.

**Comment by pivotal-jwinters** [comment]
@vito can this be solved by removing the duplicate scheduler for manually triggered builds?

**Comment by vito** [comment]
@pivotal-jwinters :turtle: Yes, I think so.

**Comment by pivotal-jwinters** [comment]
After removing the `scheduler.TriggerImmediately` from `jobserver.CreateJobBuild`, we've successfully run `ginkgo -v -untilItFails -focus='runs builds with every version'` for 300+ attempts without a failure. So I'm thinking this is fixed in fdebe4d8d5f69d89ad1cc2ab91b2906b2939eb5f.

**Comment by clarafu** [comment]
@vito seems like this is fixed with in this [commit](https://github.com/concourse/concourse/commit/fdebe4d8d5f69d89ad1cc2ab91b2906b2939eb5f)?
